src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -25,9 +25,13 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics
     /// <summary>
     /// Root type for both document and workspace diagnostic pull requests.
     /// </summary>
-    internal abstract class AbstractPullDiagnosticHandler<TDiagnosticsParams, TReport> : IRequestHandler<TDiagnosticsParams, TReport[]?>
-        where TReport : VSInternalDiagnosticReport
+    /// <typeparam name="TDiagnosticsParams">The LSP input param type</typeparam>
+    /// <typeparam name="TReport">The LSP type that is reported via IProgress</typeparam>
+    /// <typeparam name="TReturn">The LSP type that is returned on completion of the request.</typeparam>
+    internal abstract class AbstractPullDiagnosticHandler<TDiagnosticsParams, TReport, TReturn> : IRequestHandler<TDiagnosticsParams, TReturn?> where TDiagnosticsParams : IPartialResultParams<TReport[]>
     {
+        internal record PreviousResult(string PreviousResultId, TextDocumentIdentifier TextDocument);
```

Comment:
1. does thsi need to be internal?  or would protected work?
2. can this be a readonly record struct?